File: mobile/pwa/src/lib/pairing.ts
/**
 * Pairing flow for the mobile PWA (docs/MOBILE-DESIGN.md §4.2, stage S2).
 *
 * mpi://pair link → WSS to relay → pair.request → host pair.challenge →
 * Ed25519-signed pair.hello → (desktop approves) → pair.accepted + deviceToken.
 * E2E X25519/AES-GCM key material arrives in S3; the handshake itself is final.
 */
// Relative (not an alias) so node-based tests can import this module directly.
// src/lib → pwa → mobile → shared
import { makeEnvelope } from "../../../shared/protocol";
import type { DeviceIdentity } from "./device-identity";
import {
  deriveAesKeyRaw,
  decryptFrame as e2eDecrypt,
  encryptFrame as e2eEncrypt,
  importAesKey,
  x25519SharedSecretRaw,
  type E2EFrame,
} from "./e2e-crypto";
import type { RelayClient } from "./relay-client";

export interface PairingPayload {
  hostId: string;
  fingerprint?: string;
  hostPublicKeyPem?: string;
  /** Relay WSS endpoint embedded by the desktop (S3); S2 links may carry it too. */
  relayUrl?: string;
  ticket: string;
  expiresAt: number;
  protocol?: number;
  /** 桌面机器名（多设备列表用来区分主机）；旧版桌面生成的载荷里没有。 */
  hostName?: string;
}

export type PairingStage = "connecting" | "waiting-challenge" | "waiting-approval" | "approved" | "error";

/** Accepts "mpi://pair?payload=<b64url>" or a bare b64url payload. */
export function parsePairingLink(input: string): PairingPayload {
  let raw = input.trim();
  if (raw.startsWith("mpi://")) {
    const queryIndex = raw.indexOf("?");
    const params = new URLSearchParams(queryIndex >= 0 ? raw.slice(queryIndex + 1) : "");
    raw = params.get("payload") || "";
  }
  if (!raw) throw new Error("missing pairing payload");
  let json: unknown;
  try {
    const b64 = raw.replace(/-/g, "+").replace(/_/g, "/");
    const binary = atob(b64 + "=".repeat((4 - (b64.length % 4)) % 4));
    json = JSON.parse(new TextDecoder().decode(Uint8Array.from(binary, (c) => c.charCodeAt(0))));
  } catch {
    throw new Error("invalid pairing link");
  }
  if (!json || typeof json !== "object") throw new Error("invalid pairing payload");
  const value = json as Record<string, unknown>;
  if (typeof value.hostId !== "string" || !value.hostId) throw new Error("pairing payload missing hostId");
  if (typeof value.ticket !== "string" || !value.ticket) throw new Error("pairing payload missing ticket");
  return {
    hostId: value.hostId,
    fingerprint: typeof value.fingerprint === "string" ? value.fingerprint : undefined,    hostPublicKeyPem: typeof value.hostPublicKeyPem === "string" ? value.hostPublicKeyPem : undefined,
    relayUrl: typeof value.relayUrl === "string" && value.relayUrl ? value.relayUrl : undefined,
    ticket: value.ticket,
    expiresAt: Number.isSafeInteger(value.expiresAt) ? (value.expiresAt as number) : 0,
    protocol: Number.isSafeInteger(value.protocol) ? (value.protocol as number) : undefined,
    hostName: typeof value.hostName === "string" && value.hostName ? value.hostName : undefined,
  };
}

/** Resolve on the first frame matching `pred`; reject on relay/host errors / timeout. */
function awaitFrameOrError(
  client: RelayClient,
  pred: (frame: Record<string, unknown>) => boolean,
  what: string,
  timeoutMs: number,
): Promise<Record<string, unknown>> {
  return new Promise((resolve, reject) => {
    const off = client.onFrame((frame) => {
      if (frame.type === "relay.error") {
        cleanup();
        reject(new Error(`relay error: ${String(frame.code ?? "unknown")}`));
      } else if (frame.type === "error" && frame.error) {
        // Protocol-level error envelope from the host (e.g. AUTH_REQUIRED).
        const err = frame.error as Record<string, unknown>;
        cleanup();
        reject(new Error(`host error: ${String(err.code ?? "")} ${String(err.message ?? "")}`.trim()));
      } else if (pred(frame)) {
        cleanup();
        resolve(frame);
      }
    });
    const timer = setTimeout(() => {
      cleanup();
      reject(new Error(`timed out waiting for ${what}`));
    }, timeoutMs);
    function cleanup() {
      off();
      clearTimeout(timer);
    }
  });
}

async function waitForOpen(client: RelayClient, timeoutMs: number): Promise<void> {
  const deadline = Date.now() + timeoutMs;
  while (!client.isOpen()) {
    if (Date.now() > deadline) throw new Error("timed out connecting to relay");
    await new Promise((resolve) => setTimeout(resolve, 25));
  }
}

export interface PairingResult {
  deviceToken: string;
  /** Host X25519 public key (b64url) — persist it; the session key is re-derived
   * locally on every reconnect, no renegotiation needed. */
  hostX25519PubB64u: string;
}

/** Answer a pair.challenge with a signed pair.hello and wait for pair.accepted. */
async function answerChallenge(
  client: RelayClient,
  hostId: string,
  identity: DeviceIdentity,
  deviceName: string,
  ticket: string,
  acceptedTimeoutMs: number,
  onStage?: (stage: PairingStage) => void,
): Promise<PairingResult> {
  const challenge = await awaitFrameOrError(client, (f) => f.type === "pair.challenge", "pair.challenge", 20_000);
  const cp = (challenge.payload || {}) as Record<string, unknown>;
  const connectionId = typeof cp.connectionId === "string" ? cp.connectionId : "";
  const challengeValue = typeof cp.challenge === "string" ? cp.challenge : "";
  if (!connectionId || !challengeValue) throw new Error("malformed pair.challenge");

  onStage?.("waiting-approval");
  const signedText = `mpi-remote-v1|${hostId}|${connectionId}|${challengeValue}|${identity.deviceId}`;
  client.send(
    makeEnvelope("pair.hello", String(challenge.sessionId), {
      deviceId: identity.deviceId,
      deviceName,
      publicKeyPem: identity.publicKeyPem,
      signature: identity.signText(signedText),
      ticket,
      // E2E (S3): the host derives the session key from this on pair.hello.
      x25519Pub: identity.x25519PubB64u,
    }),
  );

  // Trusted devices are accepted immediately; new ones wait for the desktop user.
  const accepted = await awaitFrameOrError(client, (f) => f.type === "pair.accepted", "pair.accepted", acceptedTimeoutMs);
  const ap = (accepted.payload || {}) as Record<string, unknown>;
  const deviceToken = typeof ap.deviceToken === "string" ? ap.deviceToken : "";
  const hostX25519PubB64u = typeof ap.x25519Pub === "string" ? ap.x25519Pub : "";

  // E2E session: X25519 → HKDF-SHA256 → AES-256-GCM (§4.2). Deterministic, so a
  // reconnect re-auth (fresh pair.accepted) re-installs the same key.
  if (hostX25519PubB64u) {
    const shared = x25519SharedSecretRaw(identity.x25519PrivB64u, hostX25519PubB64u);
    const cryptoKey = await importAesKey(deriveAesKeyRaw(shared, hostId, identity.deviceId));
    // Adapter between RelayClient.FrameCrypto and the concrete E2EFrame types.
    client.setFrameCrypto({
      encrypt: async (plain) => (await e2eEncrypt(cryptoKey, plain)) as unknown as Record<string, unknown>,
      decrypt: (frame) => e2eDecrypt(cryptoKey, frame as Pick<E2EFrame, "n" | "c">),
    });
  }
  return { deviceToken, hostX25519PubB64u };
}

/** Run the full pairing handshake. Resolves with the stable deviceToken. */
export async function runPairing(
  client: RelayClient,
  payload: PairingPayload,
  identity: DeviceIdentity,
  deviceName: string,
  onStage?: (stage: PairingStage, detail?: string) => void,
): Promise<PairingResult> {
  const fail = (detail: string): never => {
    onStage?.("error", detail);
    throw new Error(detail);
  };

  client.connect();
  await waitForOpen(client, 10_000).catch(() => fail(`cannot reach relay (${client.getLastError() ?? "connection failed"})`));

  // First frame decides the socket role — must be pair.request.
  onStage?.("connecting");
  if (!client.pairRequest(payload.ticket, identity.deviceId, deviceName)) {
    return fail("failed to send pair.request");
  }

  onStage?.("waiting-challenge");
  const result = await answerChallenge(client, payload.hostId, identity, deviceName, payload.ticket, 5 * 60_000, (s) => onStage?.(s)).catch((e: Error) => fail(e.message));
  onStage?.("approved");
  return result;
}

/** Re-auth an already-trusted device after (re)connect: the relay hello triggers a
 * fresh pair.challenge from the host; answer it and wait for pair.accepted. */
export async function reauthenticate(
  client: RelayClient,
  hostId: string,
  identity: DeviceIdentity,
  deviceName: string,
): Promise<PairingResult> {
  return answerChallenge(client, hostId, identity, deviceName, "", 30_000);
}

/**
 * Keep an established session alive across mid-session socket drops: whenever the
 * socket opens again (initial connect included), redo the challenge handshake —
 * which also reinstalls E2E crypto. Attach AFTER setHelloCreds; returns cleanup.
 */
export function attachAutoReauth(
  client: RelayClient,
  hostId: string,
  identity: DeviceIdentity,
  deviceName: string,
  onReauth?: (result: PairingResult) => void,
): () => void {
  let inFlight = false;
  const off = client.onState((state) => {
    if (state !== "open" || inFlight) return;
    inFlight = true;
    reauthenticate(client, hostId, identity, deviceName)
      .then((result) => onReauth?.(result))
      .catch(() => { /* surfaced via UI error state; next open retries */ })
      .finally(() => {
        inFlight = false;
      });
  });
  return off;
}

File: mobile/app/app/src/main/java/com/mpi/app/data/Pairing.kt
package com.mpi.app.data

import com.mpi.app.protocol.Base64Url
import com.mpi.app.protocol.DeviceIdentity
import com.mpi.app.protocol.Envelope
import com.mpi.app.protocol.X25519
import com.mpi.app.protocol.deriveAesKey
import kotlinx.coroutines.ExperimentalCoroutinesApi
import kotlinx.coroutines.channels.Channel
import kotlinx.coroutines.selects.select
import kotlinx.coroutines.withTimeout
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import kotlinx.serialization.json.put

/**
 * 配对流程 —— 与 PWA 的 mobile/pwa/src/lib/pairing.ts 同构。
 *
 * 完整时序（见 docs/MOBILE-DESIGN.md §4.2）：
 *
 *   设备 → 中继(→host)  {type:"pair.request", ticket, deviceId, name}   ← 控制帧
 *   host → 设备(→中继)  envelope pair.challenge {connectionId, challenge}
 *   设备 → host         envelope pair.hello {deviceId, deviceName, publicKeyPem,
 *                                            signature, ticket, x25519Pub}
 *   host → 设备         envelope pair.accepted {deviceToken, x25519Pub}
 *   设备                 X25519(myPriv, hostPub) → HKDF → AES-256-GCM 会话密钥
 *
 * 签名文本的格式**必须与 TS 逐字节一致**，否则主机验签失败：
 *   `mpi-remote-v1|<hostId>|<connectionId>|<challenge>|<deviceId>`
 */

/** 配对阶段（与 PWA 的 PairingStage 对齐，供 UI 显示进度）。 */
enum class PairingStage {
    Connecting,
    WaitingChallenge,
    WaitingApproval,
    Approved,
}

class PairingException(message: String, cause: Throwable? = null) : Exception(message, cause)

/** 配对/重认证的产物。 */
class PairingResult(
    val deviceToken: String,
    /** 主机 X25519 公钥（b64url）。持久化它即可——会话密钥可随时本地复算，无需重新协商。 */
    val hostX25519PubB64u: String,
    /** 已派生的 AES-256-GCM 会话密钥。 */
    val aesKey: ByteArray,
)

object Pairing {
    private const val SIGNATURE_PREFIX = "mpi-remote-v1"

    /** 中继连接超时。 */
    const val CONNECT_TIMEOUT_MS = 10_000L

    /** 等 pair.challenge（中继→host→设备，多一跳）。 */
    const val CHALLENGE_TIMEOUT_MS = 20_000L

    /** 等 pair.accepted：新设备要等桌面端点「允许」，给足 5 分钟。 */
    const val ACCEPTED_TIMEOUT_MS = 5 * 60_000L

    /** 重认证时主机通常立刻回，30 秒够。 */
    const val REAUTH_ACCEPTED_TIMEOUT_MS = 30_000L

    /** 签名文本（与 TS 的 signedText 逐字节一致）。 */
    fun signedText(hostId: String, connectionId: String, challenge: String, deviceId: String): String =
        "$SIGNATURE_PREFIX|$hostId|$connectionId|$challenge|$deviceId"

    /**
     * 全新配对：连中继 → pair.request → 应答挑战 → 拿到 deviceToken 与会话密钥。
     */
    suspend fun run(
        client: RelayClient,
        hostId: String,
        ticket: String,
        identity: DeviceIdentity,
        deviceName: String,
        acceptedTimeoutMs: Long = ACCEPTED_TIMEOUT_MS,
        onStage: (PairingStage) -> Unit = {},
    ): PairingResult {
        onStage(PairingStage.Connecting)
        connectAndAwaitOpen(client)

        // 先建流再发帧：否则中继回得够快时，挑战会在注册监听（FrameStream）之前到达并丢失。
        val stream = FrameStream(client)
        try {
            // 首帧决定 socket 角色——必须是 pair.request。
            if (!client.send(ControlFrames.pairRequest(ticket, identity.deviceId, deviceName))) {
                throw PairingException("无法发送 pair.request（连接已关闭）")
            }
            onStage(PairingStage.WaitingChallenge)
            return answerChallenge(stream, hostId, identity, deviceName, ticket, acceptedTimeoutMs, onStage)
        } finally {
            stream.close()
        }
    }

    /**
     * 重认证：已配对设备（重）连上后，发 `hello` 让中继通知 host 重发挑战，
     * 应答后拿到**新的 pair.accepted**——会话密钥由静态身份重新派生，结果与首次相同。
     */
    suspend fun reauthenticate(
        client: RelayClient,
        hostId: String,
        identity: DeviceIdentity,
        deviceToken: String,
        deviceName: String,
        acceptedTimeoutMs: Long = REAUTH_ACCEPTED_TIMEOUT_MS,
        onStage: (PairingStage) -> Unit = {},
    ): PairingResult {
        onStage(PairingStage.Connecting)
        connectAndAwaitOpen(client)
        val stream = FrameStream(client)
        try {
            if (!client.send(ControlFrames.hello(identity.deviceId, deviceToken, hostId))) {
                throw PairingException("无法发送 hello（连接已关闭）")
            }
            onStage(PairingStage.WaitingChallenge)
            return answerChallenge(stream, hostId, identity, deviceName, "", acceptedTimeoutMs, onStage)
        } finally {
            stream.close()
        }
    }

    // ---- 内部 ----

    private suspend fun answerChallenge(
        stream: FrameStream,
        hostId: String,
        identity: DeviceIdentity,
        deviceName: String,
        ticket: String,
        acceptedTimeoutMs: Long,
        onStage: (PairingStage) -> Unit,
    ): PairingResult {
        val challenge = stream.await("pair.challenge", CHALLENGE_TIMEOUT_MS)
        val sessionId = challenge.str("sessionId") ?: throw PairingException("pair.challenge 缺少 sessionId")
        val payload = challenge["payload"]?.jsonObject ?: throw PairingException("pair.challenge 缺少 payload")
        val connectionId = payload.str("connectionId") ?: throw PairingException("pair.challenge 缺少 connectionId")
        val challengeValue = payload.str("challenge") ?: throw PairingException("pair.challenge 缺少 challenge")

        onStage(PairingStage.WaitingApproval)
        val signature = identity.signText(signedText(hostId, connectionId, challengeValue, identity.deviceId))
        val hello = Envelope.encode(
            Envelope.make(
                type = "pair.hello",
                sessionId = sessionId,
                payload = buildJsonObject {
                    put("deviceId", identity.deviceId)
                    put("deviceName", deviceName)
                    put("publicKeyPem", identity.publicKeyPem)
                    put("signature", signature)
                    put("ticket", ticket)
                    // E2E：主机据此派生会话密钥
                    put("x25519Pub", identity.x25519PubB64u)
                },
            ),
        )
        if (!stream.send(hello)) throw PairingException("无法发送 pair.hello（连接已关闭）")

        // 已信任设备立刻返回；新设备要等桌面端批准。
        val accepted = stream.await("pair.accepted", acceptedTimeoutMs)
        val acceptedPayload = accepted["payload"]?.jsonObject
            ?: throw PairingException("pair.accepted 缺少 payload")
        val deviceToken = acceptedPayload.str("deviceToken").orEmpty()
        if (deviceToken.isEmpty()) throw PairingException("pair.accepted 未返回 deviceToken")
        val hostX25519Pub = acceptedPayload.str("x25519Pub").orEmpty()

        val aesKey = if (hostX25519Pub.isEmpty()) {
            ByteArray(0)
        } else {
            deriveAesKey(
                X25519.sharedSecretB64u(identity.x25519PrivB64u, hostX25519Pub),
                hostId,
                identity.deviceId,
            )
        }

        onStage(PairingStage.Approved)
        return PairingResult(deviceToken = deviceToken, hostX25519PubB64u = hostX25519Pub, aesKey = aesKey)
    }

    private suspend fun connectAndAwaitOpen(client: RelayClient) {
        if (client.isOpen()) return
        client.connect()
        withTimeout(CONNECT_TIMEOUT_MS) {
            val states = Channel<RelayState>(Channel.UNLIMITED)
            val unsubscribe = client.onState { states.trySend(it) }
            try {
                if (client.isOpen()) return@withTimeout
                while (true) {
                    when (val state = states.receive()) {
                        is RelayState.Open -> return@withTimeout

                        is RelayState.Failed -> throw PairingException("无法连接中继：${state.message}")

                        // 本端主动放弃（kick 重建）不是失败：让上层走它自己的重建流程。
                        is RelayState.Closed -> if (!state.expected) {
                            throw PairingException("中继断开连接：${state.code} ${state.reason}".trim())
                        }

                        else -> Unit
                    }
                }
            } finally {
                unsubscribe()
            }
        }
    }

    /**
     * 把一路 socket 的帧收集到 Channel，供 `await(type)` 顺序取用。
     *
     * 两类「失败要快」：
     * ① 中继/主机发来的错误帧；
     * ② **连接被关闭/失败**——否则用户会干等到超时（违反了「不静默失败」的原则）。
     */
    private class FrameStream(private val client: RelayClient) {
        private val frames = Channel<JsonObject>(Channel.UNLIMITED)
        private val failures = Channel<PairingException>(Channel.UNLIMITED)

        private val unsubscribeFrame: () -> Unit = client.onFrame { raw ->
            val parsed = runCatching { Json.parseToJsonElement(raw) as? JsonObject }.getOrNull()
            if (parsed != null) frames.trySend(parsed)
        }

        private val unsubscribeState: () -> Unit = client.onState { state ->
            when (state) {
                is RelayState.Failed -> failures.trySend(PairingException("连接失败：${state.message}"))

                // 本端主动放弃（kick 重建）不当作失败上报——上层会用新 socket 重新握手。
                // 旧写法把「自己关的 1000」报成「中继关闭了连接：1000」，用户看到的是一条
                // 完全误导的错误（2026-09-26 真机事故）。
                is RelayState.Closed -> if (!state.expected) {
                    failures.trySend(
                        PairingException("中继关闭了连接：${state.code} ${state.reason}".trim()),
                    )
                }

                else -> Unit
            }
        }

        fun send(rawJson: String): Boolean = client.send(rawJson)

        @OptIn(ExperimentalCoroutinesApi::class)
        suspend fun await(type: String, timeoutMs: Long): JsonObject = withTimeout(timeoutMs) {
            var found: JsonObject? = null
            while (found == null) {
                select {
                    failures.onReceive { throw it }

                    frames.onReceive { frame ->
                        when (frame.str("type")) {
                            "relay.error" -> throw PairingException("中继错误：${frame.str("code") ?: "unknown"}")

                            "error" -> {
                                val error = frame["error"]?.jsonObject
                                val code = error?.str("code").orEmpty()
                                val message = error?.str("message").orEmpty()
                                throw PairingException("主机错误：$code $message".trim())
                            }

                            type -> found = frame
                        }
                    }
                }
            }
            found!!
        }

        fun close() {
            unsubscribeFrame()
            unsubscribeState()
            frames.close()
            failures.close()
        }
    }
}

private fun JsonObject.str(key: String): String? = this[key]?.jsonPrimitive?.contentOrNull
